dhcpv4/nclient4/client.go

```diff
@@ -469,19 +469,35 @@ func (c *Client) DiscoverOffer(ctx context.Context, modifiers ...dhcpv4.Modifier
 // Request completes the 4-way Discover-Offer-Request-Ack handshake.
 //
 // Note that modifiers will be applied *both* to Discover and Request packets.
-func (c *Client) Request(ctx context.Context, modifiers ...dhcpv4.Modifier) (lease *Lease, err error) {
+func (c *Client) Request(ctx context.Context, modifiers ...dhcpv4.Modifier) (*Lease, error) {
 	offer, err := c.DiscoverOffer(ctx, modifiers...)
 	if err != nil {
-		err = fmt.Errorf("unable to receive an offer: %w", err)
-		return
+		return nil, fmt.Errorf("unable to receive an offer: %w", err)
+	}
+
+	// TODO(chrisko): should this be unicast to the server?
+	request, err := dhcpv4.NewRequestFromOffer(offer, dhcpv4.PrependModifiers(modifiers,
+		dhcpv4.WithOption(dhcpv4.OptMaxMessageSize(MaxMessageSize)))...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create a request: %w", err)
+	}
+
+	response, err := c.TransactAckNak(ctx, request)
+	if err != nil {
+		return nil, fmt.Errorf("transaction failed: %w", err)
 	}
-	return c.RequestFromOffer(ctx, offer, modifiers...)
+
+	return &Lease{
+		Offer:        offer,
+		ACK:          response,
+		CreationTime: time.Now(),
+	}, nil
 }
 
 // ErrNak is returned if a DHCP server rejected our Request.
 type ErrNak struct {
-	Offer *dhcpv4.DHCPv4
-	Nak   *dhcpv4.DHCPv4
+	Request *dhcpv4.DHCPv4
```

Comment:
Unless I am missing something, to me it only makes sense to store what the server returns.

Reply:
I can change `ErrNak` to only store `Nak`, but that is a more significant API change than renaming `Offer` to `Request` for any library consumer that might rely on the sent request being present in `ErrNak` for failure analysis etc. If you are sufficiently confident that this hasn't been a supported use case, then I'll go ahead and do the change.

Reply:
`Offer` has now been removed.